contracts/StakingRewards.sol

```diff
@@ -0,0 +1,200 @@
+
+  function getReward() public nonReentrant updateReward(msg.sender) {
+    uint256 reward = rewards[msg.sender];
+    if (reward > 0) {
+      rewards[msg.sender] = 0;
+      rewardsToken.safeTransfer(msg.sender, reward);
+      emit RewardPaid(msg.sender, reward);
+    }
+  }
+
+  function exit() external {
+    withdraw(_balances[msg.sender]);
+    getReward();
+  }
+
+  /* ========== RESTRICTED FUNCTIONS ========== */
+
+  // Changed from 
+  // function notifyRewardAmount(uint256 reward)
+  // to 
+  // function depositReward(address, uint256 reward)
+  // for compatibility reason with IdleCDO. Added also a transferFrom to get the reward amount
+  function depositReward(address, uint256 reward) external onlyRewardsDistribution updateReward(address(0)) {
+    if (shouldTransfer) {
+      rewardsToken.safeTransferFrom(msg.sender, address(this), reward);
+    }
+
+    if (block.timestamp >= periodFinish) {
+      rewardRate = reward / rewardsDuration;
+    } else {
+      uint256 remaining = periodFinish - block.timestamp;
+      uint256 leftover = remaining * rewardRate;
+      rewardRate = (reward + leftover) / rewardsDuration;
+    }
+
+    // Ensure the provided reward amount is not more than the balance in the contract.
+    // This keeps the reward rate in the right range, preventing overflows due to
+    // very high values of rewardRate in the earned and rewardsPerToken functions;
+    // Reward + leftover must be less than 2^256 / 10^18 to avoid overflow.
+    uint balance = rewardsToken.balanceOf(address(this));
+    require(rewardRate <= balance / rewardsDuration, "Provided reward too high");
+
+    lastUpdateTime = block.timestamp;
+    periodFinish = block.timestamp + rewardsDuration;
+
+    emit RewardAdded(reward);
+  }
+
+  // Added to support recovering LP Rewards from other systems such as BAL to be distributed to holders
+  function recoverERC20(address tokenAddress, uint256 tokenAmount) external onlyOwner {
+    require(tokenAddress != address(stakingToken), "Cannot withdraw the staking token");
+    IERC20(tokenAddress).safeTransfer(owner(), tokenAmount);
+    emit Recovered(tokenAddress, tokenAmount);
+  }
+
+  function setRewardsDuration(uint256 _rewardsDuration) external onlyOwner {
+    require(
+      block.timestamp > periodFinish,
+      "Previous rewards period must be complete before changing the duration for the new period"
+    );
+    rewardsDuration = _rewardsDuration;
+    emit RewardsDurationUpdated(rewardsDuration);
+  }
+
+  function setRewardsDistribution(address _rewardsDistribution) external onlyOwner {
+    rewardsDistribution = _rewardsDistribution;
+  }
+
+  /* ========== MODIFIERS ========== */
+
+  modifier updateReward(address account) {
+    rewardPerTokenStored = rewardPerToken();
+    lastUpdateTime = lastTimeRewardApplicable();
+    if (account != address(0)) {
+        rewards[account] = earned(account);
+        userRewardPerTokenPaid[account] = rewardPerTokenStored;
+    }
+    _;
+  }
```

Comment:
Function modifiers can be inefficient, I think.
> When you add a function modifier, the code of that function is picked up and put in the function modifier in place of the _ symbol. This can also be understood as ‘The function modifiers are inlined”. If the same code is inlined multiple times, it adds up in size.

> Internal functions, on the other hand, are not inlined but called as separate functions. This means they are very slightly more expensive in run time but save a lot of redundant bytecode in deployment.

ref: https://blog.polymath.network/solidity-tips-and-tricks-to-save-gas-and-reduce-bytecode-size-c44580b218e6

So, I suggest :
```suggestion solidity
  modifier updateReward(address account) {
    _updateReward(account);
    _;
  }
```

Reply:
I agree with you as the code is basically duplicated across methods, but the idea here was to have the implementation as close as possible to the original one (but still compatible to IdleCDO of course)